src/View/SSTemplateEngine.php

```diff
@@ -0,0 +1,467 @@
+    }
+
+    /**
+     * Get the appropriate template to use for the named sub-template, or null if none are appropriate
+     */
+    protected function getSubtemplateFor(string $subtemplate): ?array
+    {
+        // Get explicit subtemplate name
+        if (isset($this->subTemplates[$subtemplate])) {
+            return $this->subTemplates[$subtemplate];
+        }
+
+        // Don't apply sub-templates if type is already specified (e.g. 'Includes')
+        if (isset($this->templateCandidates['type'])) {
+            return null;
+        }
+
+        // Filter out any other typed templates as we can only add, not change type
+        $templates = array_filter(
+            (array) $this->templateCandidates,
+            function ($template) {
+                return !isset($template['type']);
+            }
+        );
+        if (empty($templates)) {
+            return null;
+        }
+
+        // Set type to subtemplate
+        $templates['type'] = $subtemplate;
+        return $templates;
+    }
+
+    /**
+     * Parse given template contents
+     *
+     * @param string $content The template contents
+     * @param string $template The template file name
+     */
+    protected function parseTemplateContent(string $content, string $template = ""): string
+    {
+        return $this->getParser()->compileString(
+            $content,
+            $template,
+            Director::isDev() && SSViewer::config()->uninherited('source_file_comments')
+        );
+    }
+
+    /**
+     * Attempts to find possible candidate templates from a set of template
+     * names from modules, current theme directory and finally the application
+     * folder.
+     *
+     * The template names can be passed in as plain strings, or be in the
+     * format "type/name", where type is the type of template to search for
+     * (e.g. Includes, Layout).
+     *
+     * The results of this method will be cached for future use.
+     *
+     * @param string|array $template Template name, or template spec in array format with the keys
+     * 'type' (type string) and 'templates' (template hierarchy in order of precedence).
+     * If 'templates' is omitted then any other item in the array will be treated as the template
+     * list, or list of templates each in the array spec given.
+     * Templates with an .ss extension will be treated as file paths, and will bypass
+     * theme-coupled resolution.
+     * @param array $themes List of themes to use to resolve themes. Defaults to {@see SSViewer::get_themes()}
+     * @return string Absolute path to resolved template file, or null if not resolved.
+     * File location will be in the format themes/<theme>/templates/<directories>/<type>/<basename>.ss
+     * Note that type (e.g. 'Layout') is not the root level directory under 'templates'.
+     * Returns null if no template was found.
+     */
+    private function findTemplate(string|array $template, array $themes = []): ?string
```

Comment:
This looks like the sort of code that should be abstracted and put on TemplateEngine.php

Reply:
Same goes for ... probably most of the methods on this class? Seems like TemplateEngine.php should be an abstract class rather than an Interface

Reply:
> This looks like the sort of code that should be abstracted and put on TemplateEngine.php

It's private, there's no reason to put it on the interface. The intention is to dissuade people from trying to deal with exact absolute template file paths and instead reference template candidates (without file extensions!) so that when template engines get swapped out, we're not trying to feed a ss template file to a twig engine or a moustache template file to a blade engine.

> Same goes for ... probably most of the methods on this class? Seems like TemplateEngine.php should be an abstract class rather than an Interface

Most if not _all_ of the functionality in `SSTemplateEngine` is specific to ss template files and the ss template syntax. It wouldn't be useful at all for other template engines.